# Datos:
* Estudiante: Juan Jose Valverde Campos
* Carnet: B47200
* Laboratorio 3 - LR
* Profesor: Pablo Sauma Chacón

Para este laboratorio se trabajará con el set de datos: fish_perch.csv que consiste de datos estrictamente numéricos. El objetivo de este set de datos es predecir el peso weight de cada pez en función de sus otros atributos: Length1, Length2, Length3, Width y Height. 

In [2]:
import pandas as pd 
import numpy as np
from LR import LR # Clase implementada.
from sklearn.model_selection import PredefinedSplit, train_test_split

df =  pd.read_csv('fish_perch.csv',delimiter=",",decimal=".")
Y = df.pop("Weight")
myLr = LR()
def score(y_true, y_predict):
    '''
    Funcion de R cuadrado
    '''
    SSR = np.sum(np.power((y_true - y_predict),2)) # Sum of squared regression
    SST = np.sum(np.power((y_true - np.mean(y_true)),2)) # Sum of squared total 
    return 1-SSR / SST

### Para cada prueba calcule e imprima el error R2, realice la prueba con múltiples combinaciones de parámetros, intentando mejorar la estimación

Nota: En este caso se presentan combinaciones con parámetros que ya previamente se detectó que dan buenos resultados, a excpeción de la prueba 0, que corresponde a la inicialización de todos los hiperparámetros dados por defecto.

In [3]:
X_train, X_test, y_train, y_test = train_test_split(df, Y, train_size=0.75, test_size=0.25, random_state=21)

# Prueba 0
myLr.fit(X_train, y_train)
y_pred = myLr.predict(X_test)
print("Prueba 0. ",score(y_test,y_pred))
# Prueba 1
myLr.fit(X_train, y_train,max_epochs=100000, threshold=0.000001,learning_rate=0.0001,momentum=1e-2, decay=1e-5, error = 'mse', regularization='l2', lamda=1e-3)
y_pred = myLr.predict(X_test)
print("Prueba 1. ",score(y_test,y_pred))
# Prueba 2
myLr.fit(X_train, y_train,max_epochs=10000000, threshold=0.0000001,learning_rate=0.0001,momentum=1e-2, decay=0, error = 'mse', regularization='elastic-net', lamda=1e-3)
y_pred = myLr.predict(X_test)
print("Prueba 2. ",score(y_test,y_pred))
# Prueba 3
myLr.fit(X_train, y_train,max_epochs=10000000, threshold=0.0000001,learning_rate=0.0001,momentum=1e-3, decay=0, error = 'mse', regularization='elastic-net', lamda=1e-3)
y_pred = myLr.predict(X_test)
print("Prueba 3. ",score(y_test,y_pred))
# Prueba 4
myLr.fit(X_train, y_train,max_epochs=100000, threshold=0.000000001,learning_rate=1e-2,momentum=0.99, decay=1e-5, error = 'mae', regularization='l1', lamda=1e-4)
y_pred = myLr.predict(X_test)
print("Prueba 4. ",score(y_test,y_pred))
# Prueba 5
myLr.fit(X_train, y_train,max_epochs=1000000, threshold=0.000000001,learning_rate=1e-1,momentum=0.99, decay=1e-6, error = 'mae', regularization='l1', lamda=1e-2)
y_pred = myLr.predict(X_test)
print("Prueba 5. ",score(y_test,y_pred))
# Prueba 6
myLr.fit(X_train, y_train,max_epochs=1000000, threshold=0.000000001,learning_rate=1e-1,momentum=0.50, decay=1e-6, error = 'mae', regularization='l1', lamda=1e-2)
y_pred = myLr.predict(X_test)
print("Prueba 6. ",score(y_test,y_pred))

Prueba 0.  -1.833386347214416e+130
Prueba 1.  0.8783435198809615
Prueba 2.  0.9114086725602027
Prueba 3.  0.9113760047857967
Prueba 4.  0.8265941633208307
Prueba 5.  0.8458441123215248
Prueba 6.  0.9341589552760945
Prueba 7.  0.9148094063118365


### ¿Cuál fue la combinación de parámetros que le proveyó el mejor resultado?
En este caso la combinación que dio mejores resultados corresponde:
* A la combinación de la prueba 2 con un R2 reportado de 0.9114, para el caso de uso de función de error MSE.
* A la combinación de la prueba 6 con un R2 reportado de 0.9341, para el caso de uso de función de error MAE.

Dado lo anterior parece que el mejor modelo es el de la prueba 6 con un R2 reportado de 0.9341, no obstante para el siguiente segmento se utilizará también la combinación de la prueba 2, para ver que sucede según la semilla que se seleccione.

# ¿Qué pasa si utiliza esa misma combinación pero cambia la semilla del train_test_split?  Pruebe con varias semillas
* Nota: Se utiliza tres tipos de semilla distinta para cada uno de los casos [42,10,5]
1. Los resultados muestran que según el tipo de semilla se tienen resultados distintos, donde por ejemplo según el tipo de semilla la función de error MSE supera a la función de error de MAE, contrario a lo que se tenía previamente (Donde hay que recordar que estos son otro más de los hiperparámetros del modelo).
2. Además se observa que incluso el resultado en solo una de las semillas continúa con resultados por sobre el 90% en el R cuadrado, semilla = 5, como se tenían previamente. Lo que demuestra el impacto de la semilla seleccionada.

In [4]:
randoms = [42,10,5]

for random in randoms : 
    X_train, X_test, y_train, y_test = train_test_split(df, Y, train_size=0.75, test_size=0.25, random_state=random)
    myLr.fit(X_train, y_train,max_epochs=10000000, threshold=0.0000001,learning_rate=0.0001,momentum=1e-3, decay=0, error = 'mse', regularization='elastic-net', lamda=1e-3)
    y_pred = myLr.predict(X_test)
    print("Prueba de valor random = ", random, " con MSE: ",score(y_test,y_pred))
    myLr.fit(X_train, y_train,max_epochs=1000000, threshold=0.000000001,learning_rate=1e-1,momentum=0.50, decay=1e-6, error = 'mae', regularization='l1', lamda=1e-2)
    y_pred = myLr.predict(X_test)
    print("Prueba de valor random = ", random, " con MAE: ",score(y_test,y_pred))

Prueba de valor random =  42  con MSE:  0.8843046724780531
Prueba de valor random =  42  con MAE:  0.8298235445615041
Prueba de valor random =  10  con MSE:  0.7927893170495759
Prueba de valor random =  10  con MAE:  0.8045185124638912
Prueba de valor random =  5  con MSE:  0.950067726733323
Prueba de valor random =  5  con MAE:  0.9167714874795835


### Si pasa algo inusual: ¿Por qué cree que pasa esto?
Esto sucede por la forma que se realiza la separación de datos, lo cual provocan que datos de entrenamiento y datos de testing sean distintos según el tipo de semilla elegido. Esto provoca que los hiperparámetros del modelo elegido como el "mejor" pueda estar presentando un sesgo para un tipo de semilla específico (Un tipo de separación específico de entrenamiento y testing), lo que puede resolverse utilizando lo visto en clase de tener un tercer conjunto de datos de validación.